File: global-hooks/framework/guardrails/enforce_orchestrate.py
#!/usr/bin/env python3
"""
Enforce /orchestrate Skill Invocation - UserPromptSubmit Hook
=============================================================

Detects when the user types /orchestrate (or asks to orchestrate) and injects
a BLOCKING system reminder that forces the Skill tool to be invoked before
any other action. This prevents Claude from ignoring /orchestrate and doing
work single-threaded in the main agent.

Trigger: UserPromptSubmit
Exit: Always 0 (never blocks)
Output: additionalContext with mandatory Skill invocation reminder
"""

import json
import os
import re
import sys
from pathlib import Path


# ---------------------------------------------------------------------------
# Detection patterns
# ---------------------------------------------------------------------------

# Direct slash command
SLASH_ORCHESTRATE = re.compile(r"(?:^|\s)/orchestrate\b", re.IGNORECASE)

# Natural language requests that should also trigger orchestration
ORCHESTRATE_INTENT = re.compile(
    r"\b(?:orchestrate\b|run orchestrat\w*|use orchestrat\w*|spawn orchestrat\w*|"
    r"parallel agents?\b|multi.?agent\b|spawn.*team\b)",
    re.IGNORECASE,
)


def needs_orchestrate_enforcement(prompt: str) -> bool:
    """Return True if the prompt contains /orchestrate or orchestration intent."""
    stripped = prompt.strip()
    if not stripped:
        return False

    # Direct slash command — always enforce
    if SLASH_ORCHESTRATE.search(stripped):
        return True

    # Natural language — only if clearly requesting orchestration
    if ORCHESTRATE_INTENT.search(stripped):
        return True

    return False


# ---------------------------------------------------------------------------
# Enforcement reminder
# ---------------------------------------------------------------------------

ORCHESTRATE_ENFORCEMENT = """\
[ORCHESTRATE ENFORCEMENT — BLOCKING REQUIREMENT]
The user has requested /orchestrate. You MUST:

1. IMMEDIATELY call: Skill(skill="orchestrate", args="<user's full message>")
2. Do NOT read files, do NOT write code, do NOT research — the Skill tool FIRST
3. Do NOT treat /orchestrate as decorative text — it is a COMMAND
4. The skill will load the orchestrator protocol — YOU are the orchestrator
5. You coordinate by spawning parallel Agent() calls — never do work yourself

SEQUENCE: Skill("orchestrate") → protocol loads → YOU spawn parallel agents (researchers, builders, validators)
VIOLATION: Any Read/Edit/Bash call instead of spawning agents = failure"""

# Marker file for orchestrator_tool_guard.py
ORCH_GUARD_MARKER = "/tmp/caf_orch_guard.marker"
ORCH_DEPTH_FILE = "/tmp/caf_orch_depth"


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    try:
        # CAF_MODE check
        _framework_dir = Path(__file__).parent.parent
        if str(_framework_dir) not in sys.path:
            sys.path.insert(0, str(_framework_dir))
        try:
            from caf_mode import should_run
            if not should_run("enforce_orchestrate"):
                sys.exit(0)
        except ImportError:
            pass

        # Dry-run mode: print to stderr instead of stdout so enforcement
        # is visible in logs but does NOT inject into Claude's context.
        dry_run = os.environ.get("CAF_DRY_RUN") == "1"

        input_data = json.load(sys.stdin)
        prompt = input_data.get("prompt", "")

        if needs_orchestrate_enforcement(prompt):
            output = {
                "hookSpecificOutput": {
                    "hookEventName": "UserPromptSubmit",
                    "additionalContext": ORCHESTRATE_ENFORCEMENT,
                }
            }
            if dry_run:
                print(
                    f"[DRY-RUN] enforce_orchestrate would inject:\n{json.dumps(output, indent=2)}",
                    file=sys.stderr,
                )
            else:
                print(json.dumps(output))

            # Note: marker/depth files are managed by orch_depth_tracker.py
            # The guard hook blocks orchestrator SUBAGENTS from using
            # Read/Grep/Glob/Edit (forces delegation). The root agent
            # is NOT blocked — it coordinates by spawning agents directly.

        sys.exit(0)

    except Exception:
        # Never block
        sys.exit(0)


if __name__ == "__main__":
    try:
        main()
    except Exception:
        pass

File: caf-hooks/src/hooks/enforce_orchestrate.rs
/// Enforce Orchestrate — UserPromptSubmit hook.
///
/// Python equivalent: global-hooks/framework/guardrails/enforce_orchestrate.py (132 LOC)
///
/// Behavior:
/// - Detect if user prompt contains /orchestrate or orchestration intent
/// - If so, inject a blocking enforcement reminder into additionalContext
/// - Always exits 0
use regex::Regex;
use std::sync::OnceLock;

use crate::io::{read_stdin_value, write_output};
use crate::types::HookOutput;

const ORCHESTRATE_ENFORCEMENT_SHORT: &str = "\
[ORCHESTRATE] Call Skill(skill=\"orchestrate\") FIRST — do NOT read/write/run before the skill loads.
Spawn parallel Agent() calls; never do the work yourself.";

const ORCHESTRATE_ENFORCEMENT: &str = "\
[ORCHESTRATE ENFORCEMENT — BLOCKING REQUIREMENT]
The user has requested /orchestrate. You MUST:

1. IMMEDIATELY call: Skill(skill=\"orchestrate\", args=\"<user's full message>\")
2. Do NOT read files, do NOT write code, do NOT research — the Skill tool FIRST
3. Do NOT treat /orchestrate as decorative text — it is a COMMAND
4. The skill will load the orchestrator protocol — YOU are the orchestrator
5. You coordinate by spawning parallel Agent() calls — never do work yourself

SEQUENCE: Skill(\"orchestrate\") → protocol loads → YOU spawn parallel agents (researchers, builders, validators)
VIOLATION: Any Read/Edit/Bash call instead of spawning agents = failure";

static SLASH_ORCHESTRATE: OnceLock<Regex> = OnceLock::new();
static ORCHESTRATE_INTENT: OnceLock<Regex> = OnceLock::new();

fn slash_orchestrate() -> &'static Regex {
    SLASH_ORCHESTRATE.get_or_init(|| {
        Regex::new(r"(?i)(?:^|\s)/orchestrate\b").expect("valid regex")
    })
}

fn orchestrate_intent() -> &'static Regex {
    ORCHESTRATE_INTENT.get_or_init(|| {
        Regex::new(r"(?i)\b(?:orchestrate\b|run orchestrat\w*|use orchestrat\w*|spawn orchestrat\w*|parallel agents?\b|multi.?agent\b|spawn.*team\b)").expect("valid regex")
    })
}

/// Returns true if /tmp/caf_caddy_result.json exists and contains strategy == "orchestrate".
fn caddy_already_says_orchestrate() -> bool {
    let path = "/tmp/caf_caddy_result.json";
    if let Ok(contents) = std::fs::read_to_string(path) {
        if let Ok(v) = serde_json::from_str::<serde_json::Value>(&contents) {
            return v.get("strategy").and_then(|s| s.as_str()) == Some("orchestrate");
        }
    }
    false
}

fn needs_orchestrate_enforcement(prompt: &str) -> bool {
    let stripped = prompt.trim();
    if stripped.is_empty() {
        return false;
    }
    if slash_orchestrate().is_match(stripped) {
        return true;
    }
    if orchestrate_intent().is_match(stripped) {
        return true;
    }
    false
}

pub fn run() {
    let data = read_stdin_value();

    let prompt = data
        .get("prompt")
        .and_then(|v| v.as_str())
        .unwrap_or("");

    if needs_orchestrate_enforcement(prompt) {
        let message = if caddy_already_says_orchestrate() {
            ORCHESTRATE_ENFORCEMENT_SHORT
        } else {
            ORCHESTRATE_ENFORCEMENT
        };
        let output = HookOutput::inject_context("UserPromptSubmit", message);
        write_output(&output);
    } else {
        write_output(&serde_json::json!({}));
    }
}
